sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -133,37 +133,50 @@
   private final ValueProvider<Integer> numShardsProvider;
   private final boolean windowedWrites;
   private int maxNumWritersPerBundle;
+  // If set,this is used instead of DynamicDestinations.getSideInputs. This is used by the Beam
+  // translation code.
+  private final List<PCollectionView<?>> overriddenSideInputs;
+
+  private List<PCollectionView<?>> getSideInputs() {
+    return (overriddenSideInputs != null)
+        ? overriddenSideInputs
+        : getSink().getDynamicDestinations().getSideInputs();
```

Comment:
So this is not different from every other place, it's exactly what BigQueryIO does :)

The background here: unlike ParDo, which users use directly, users generally don't interact directly with WriteFiles. The way users configure dynamic inputs in WriteFiles is by passing in an implementation of DynamicDestinations. Since the whole reason for the side inputs is that they need to be accessed by other function overrides in DynamicDestinations (often getSchema) and nowhere else, it made sense for the user to have their DynamicDestinations object declare what side inputs it needed.